projects/nginx/fuzz/http_request_fuzzer.c

```diff
@@ -0,0 +1,321 @@
+  memcpy(buf, request.data, size);
+  request.data += size;
+  request.data_len -= size;
+  return size;
+}
+
+// Feed fuzzing input for the reply from upstream
+static ssize_t reply_recv_handler(ngx_connection_t *c, u_char *buf, size_t size) {
+  req_reply = (ngx_http_request_t *)(c->data);
+  if(!cln_added){ // add cleanup so that we know whether everything is cleanup correctly
+    cln_added=1;
+    cln_new.handler = cleanup_reply;
+    cln_new.next = req_reply->cleanup;
+    cln_new.data = NULL;
+    req_reply->cleanup = &cln_new;
+  }
+  upstream = req_reply->upstream;
+
+  if (reply.data_len < size)
+    size = reply.data_len;
+  memcpy(buf, reply.data, size);
+  reply.data += size;
+  reply.data_len -= size; 
+  if (size == 0)
+    c->read->ready = 0;
+  return size;
+}
+
+static ngx_int_t add_event(ngx_event_t *ev, ngx_int_t event, ngx_uint_t flags){
+  return NGX_OK;
+}
+
+static ngx_int_t init_event(ngx_cycle_t *cycle, ngx_msec_t timer){
+  return NGX_OK;
+}
+
+
+// Used when sending data, do nothing
+static ngx_chain_t *send_chain(ngx_connection_t *c, ngx_chain_t *in, off_t limit) {
+  c->read->ready=1;
+  c->recv = reply_recv_handler;
+  return in->next;
+}
+
+
+// Create a base state for Nginx without starting the server
+int InitializeNginx(void) {
+  ngx_log_t *log;
+  ngx_cycle_t init_cycle;
+
+  if( access("nginx.sock", F_OK ) != -1 ) {
+    remove("nginx.sock");
+  }
+
+  ngx_debug_init();
+  ngx_strerror_init();
+  ngx_time_init();
+#if (NGX_PCRE)
+  ngx_regex_init();
+#endif
+
+  // Just output logs to stderr
+  ngx_log.file = &ngx_log_file;
+  ngx_log.log_level = NGX_LOG_EMERG;
+  ngx_log_file.fd = ngx_stderr;
+  log = &ngx_log;
+
+#if (NGX_OPENSSL)
+  ngx_ssl_init(log);
+#endif
+  ngx_memzero(&init_cycle, sizeof(ngx_cycle_t));
+  init_cycle.log = log;
+  ngx_cycle = &init_cycle;
+
+  init_cycle.pool = ngx_create_pool(1024, log);
+
+  // Set custom argv/argc
+  my_argv[0] = arg1;
+  my_argv[1] = NULL;
+  ngx_argv = ngx_os_argv = my_argv;
+  ngx_argc = 0;
+
+  // Weird trick to free a leaking buffer always caught by ASAN
+  // We basically let ngx overwrite the environment variable, free the leak and
+  // restore the environment as before.
+  char *env_before = environ[0];
+  environ[0] = my_argv[0] + 1;
+  ngx_os_init(log);
+  free(environ[0]);
+  environ[0] = env_before;
+
+  ngx_crc32_table_init();
+  ngx_preinit_modules();
+
+  FILE *fptr = fopen(config_file, "w");
+  fprintf(fptr, configuration);
+  fclose(fptr);
+  init_cycle.conf_file.len = strlen(config_file);
+  init_cycle.conf_file.data = (unsigned char *) config_file;
+
+  cycle = ngx_init_cycle(&init_cycle);
+
+  ngx_os_status(cycle->log);
+  ngx_cycle = cycle;
+
+  ngx_event_actions.add = add_event;
+  ngx_event_actions.init = init_event;
+  ngx_io.send_chain = send_chain;
+  ngx_event_flags =1;
+  ngx_event_timer_init(cycle->log);
+  return 0;
+}
+
+
+void *invalid_call(void) {
+  return NULL;
+}
+
+
+int LLVMFuzzerInitialize(int *argc, char ***argv){
```

Comment:
it's better to use static initialization instead of `LLVMFuzzerInitialize`, unless you actually need access to `argc / argv`, see more on this in https://github.com/google/fuzzing/blob/master/docs/good-fuzz-target.md#one-time-initialization and http://llvm.org/docs/LibFuzzer.html#startup-initialization